### PR TITLE
Enhancements for SynDiff Compatibility with Neuropoly's Data

### PR DESCRIPTION
**Pull request by ArthurBoschet**
### Overview

This PR introduces a series of updates aimed at enhancing the SynDiff model to ensure compatibility with Neuropoly's datasets and to improve its overall functionality and ease of use. The changes ensure a more robust use of unsupervised medical image translation.

#### Dataset Class for Neuropoly's Data
- A new dataset class has been implemented to handle hdf5 files containing Neuropoly's data effectively. This class is designed to accommodate datasets with varying numbers of images across modalities, a significant improvement over the previous limitation of one-to-one image pairings across epochs. This enhancement not only makes the model compatible with Neuropoly's data but also advances its capability in unsupervised translation tasks by allowing for more flexible image pairings, thereby reflecting the real-world advantages of unsupervised image translation methods more accurately.

#### EMA Compatibility with New PyTorch Optimizers
- A bug fix has been applied to the Exponential Moving Average (EMA) implementation to ensure compatibility with newer versions of PyTorch optimizers. This fix addresses potential compatibility issues and enhances the model's stability and performance across different versions of PyTorch.

#### Shuffling of Modality Pairings
- To improve the model's generalization ability, the training script now shuffles modality pairings at the start of each epoch. This approach introduces variability in the training process, which is beneficial for the model's ability to generalize from the training data to unseen data.

#### Environment Configuration
- An `environment.yaml` file has been added to streamline the setup process for new users. This file contains all the necessary dependencies, making it easier to create a conda environment that is ready for running SynDiff with the intended datasets.

#### Updated README
- The README.md has been updated to provide clear and concise instructions on how to utilize the repository, especially in the context of Neuropoly's data. This update ensures that users can easily navigate the setup process and understand the model's capabilities and requirements.

#### Gitignore Update
- The local folder has been added to the `.gitignore` file to prevent unnecessary files from being tracked by Git, ensuring a cleaner and more manageable codebase.

### CODE REVIEW

File: utils/EMA.py
Comment (hermancollin) [inline utils/EMA.py:1]: What torch version compatiblity issue is adressed here? e.g. which version were made compatible with this change?
Maybe a comment could be added there for future reference.

File: dataset.py
Comment (hermancollin) [inline dataset.py:1]: Great modification. 
Any idea why there was no shuffling done on the pairs in the original work? Is it possible their data was massive compared to ours? It would make sense if they trained on MRI slices because this can generate a lot of data from a few subjects.

**Review by hermancollin — APPROVED** [review]

File: utils/EMA.py [hunk 1]
```diff
@@ -21,6 +21,16 @@ def __init__(self, opt, ema_decay):
         self.optimizer = opt
         self.state = opt.state
         self.param_groups = opt.param_groups
+
```
Comment (hermancollin) [inline utils/EMA.py]:
```suggestion
        # these lines were changed to accomodate a torch version >= 2.2.0
```

File: dataset.py
Comment (hermancollin) [reply dataset.py:1]: A physical discussion confirmed this and also SynDiff used a slice-wise pairing between modalities (T1, T2) in the BRATS dataset.
